src/Persistence/Array_.php

```diff
@@ -225,39 +230,37 @@ public function generateNewId(Model $model)
 
         switch ($type) {
             case 'integer':
-                $ids = $model->id_field ? array_keys($this->data[$model->table]) : [count($this->data[$model->table])];
-
-                $id = $ids ? max($ids) + 1 : 1;
+                $nextId = ($this->maxSeenIdByTable[$model->table] ?? 0) + 1;
+                $this->maxSeenIdByTable[$model->table] = $nextId;
 
                 break;
             case 'string':
-                $id = uniqid();
+                $nextId = uniqid();
 
                 break;
             default:
                 throw (new Exception('Unsupported id field type. Array supports type=integer or type=string only'))
                     ->addMoreInfo('type', $type);
         }
 
-        $this->lastInsertIds[$model->table] = $id;
-        $this->lastInsertIds['$'] = $id;
+        $this->lastInsertIdByTable[$model->table] = $nextId;
+        $this->lastInsertIdTable = $model->table;
 
-        return $id;
+        return $nextId;
     }
 
     /**
      * Last ID inserted.
-     * Last inserted ID for any table is stored under '$' key.
      *
      * @return mixed
      */
     public function lastInsertId(Model $model = null)
     {
         if ($model) {
-            return $this->lastInsertIds[$model->table] ?? null;
+            return $this->lastInsertIdByTable[$model->table] ?? null;
         }
 
-        return $this->lastInsertIds['$'] ?? null;
+        return $this->lastInsertIdByTable[$this->lastInsertIdTable] ?? null;
```

Comment:
what if `lastInsertIdTable` is null (no inserts was done yet) ?

Reply:
![image](https://user-images.githubusercontent.com/2228672/115585762-31f37900-a2cc-11eb-8eee-ea679b54dbe6.png)